src/app/ReadClient.cpp

```diff
@@ -68,20 +68,29 @@ void ReadClient::ClearActiveSubscriptionState()
 
 void ReadClient::StopResubscription()
 {
-
     CancelLivenessCheckTimer();
     CancelResubscribeTimer();
-    mpCallback.OnDeallocatePaths(std::move(mReadPrepareParams));
+
+    // Only deallocate the paths if they are not already deallocated.
+    if (mReadPrepareParams.mpAttributePathParamsList != nullptr || mReadPrepareParams.mpEventPathParamsList != nullptr ||
+        mReadPrepareParams.mpDataVersionFilterList != nullptr)
+    {
+        mpCallback.OnDeallocatePaths(std::move(mReadPrepareParams));
+        // Make sure we will never try to free those pointers again.
+        mReadPrepareParams.mpAttributePathParamsList    = nullptr;
+        mReadPrepareParams.mAttributePathParamsListSize = 0;
+        mReadPrepareParams.mpEventPathParamsList        = nullptr;
+        mReadPrepareParams.mEventPathParamsListSize     = 0;
+        mReadPrepareParams.mpDataVersionFilterList      = nullptr;
+        mReadPrepareParams.mDataVersionFilterListSize   = 0;
+    }
 }
 
 ReadClient::~ReadClient()
 {
+    Close(CHIP_NO_ERROR, /* allowResubscription = */ false, /* allowOnDone = */ false);
```

Comment:
Calling `Close` here is a bit odd. If we arrived at the destructor here as a side-effect of a previous `Close`, it's going to call it twice. That isn't in theory, a problem per say, but I tend to think we should avoid such double calls if we can in the design here if possible.

The only thing you really need out of this `Close` is the call to `StopResubscription` really. So, why not move `StopSubscription` exclusively into the destructor, and then avoid this call entirely?

Reply:
> The only thing you really need out of this Close is the call to StopResubscription really. So, why not move StopSubscription exclusively into the destructor, and then avoid this call entirely?

I actually went back and forth about this some: if we add more logic to Close() it's easy to forget to do it in the destructor....  But yes, that is a much smaller and simpler change and I can absolutely do that.

Reply:
This was merged without resolving this. Please fix this in a subsequent PR.

Reply:
https://github.com/project-chip/connectedhomeip/issues/22352 and https://github.com/project-chip/connectedhomeip/pull/22353